### ISSUE TITLE
HttpNetworkSender TCP Connection Issue

### ISSUE THREAD

**Issue by amikapoor** [NLog/NLog#3960]
In `HttpNetworkSender.cs`, is there any reason why `HttpWebRequest` is used over a static `HttpClient`?

As far as I know, `HttpClient` is much more performant than `HttpWebRequest`, which is only being kept around for compatibility reasons.

The reason I am asking is because in an Azure Functions app where NLog is used to log to Loggly, I have noticed some issues. When logging is turned on, the app starts using thousands of TCP connections causing it to shutdown. Without logging, there are no issues.

Furthermore, if I use the [NLog.Targets.Loggly extension](https://github.com/neutmute/nlog-targets-loggly), I see no TCP connection issues because it uses a static `HttpClient`. While using this instead of the `Network Target` is a solution, it would be great if an extension wasn't needed because it has other different issues.

 I have deployed a custom Azure Functions app where I can input an integer and that many log statements will be logged through the `Network Target`. I have reproduced the issue this way.

**Comment by 304NotModified** [comment]
Good question, I see it's supported for .NET Standard, but not .NET Framework 3,5-4,5 out of the box

![image](https://user-images.githubusercontent.com/5808377/81982172-aee49000-9631-11ea-868c-a2b28307e929.png)

Maybe we could use the HttpClient indeed.

@amikapoor which platform are you using?

**Comment by amikapoor** [comment]
@304NotModified Thanks for the response.

I'm using netcore2.1.

Do you think it would also be keep it as one target but just have the new code for the newer platforms?

**Comment by snakefoot** [comment]
It would make sense to have a bulk-copy HttpClient-target in the NLog-core.

The NLog-NetworkTarget is not perfect for HttpCleint, as it is highly optimized for Tcp-Stream, where you can make multiple individual writes using the same Tcp-Connections (Not opening/closing the connection for each requerst).

If there should be a HttpClient-Target then it should be optimized to batch multiple LogEvents into single Http-request. Much like how NLog FileTarget operates to optimize for KeepFileOpen=false.

I guess inspiration for using HttpClient can be found here:

https://github.com/DarekDan/NLog.Targets.HTTP (Nuget: [NLog.Targets.Http](https://www.nuget.org/packages/NLog.Targets.HTTP))

**Comment by amikapoor** [comment]
I have ran some tests on my deployed function app and attached are the results.

The function app is simple. I give it a number and it logs that many log statements in a for loop. For all these tests, I gave an input of 40,000.

From the testing, we can see that the Network target when using Http performs the worst of the 3. The TCP connections and SNAT port allocations go into the thousands, which eventaully will cause the app to shutdown.

loggly-csharp 4.6.1.82, which is the version prior to my changes, performs better on the TCP side because it reuses the same HttpClient, but it still causes way more SNAT port allocations than necessary because it was unnecessarily closing each connection after each request.

loggly-csharp 4.6.1.90 is the latest version which has the changes I made. We can see here that TCP connections and SNAT port allocations are very low.

TLDR;
|      | TCP Ports | SNAT Ports |
| ---| -----------|-------------|
|NLog Network Target Http| ~3,000|~3,000|
|loggly-csharp 4.6.1.82|~30|~9,000|
|loggly-csharp 4.6.1.90|~30|~300|

If the Network target is going to have the `HttpNetworkSender`, I think it should be optimized so it doesn't cause these types of issues.

![Logging Comparison](https://user-images.githubusercontent.com/29603337/82158053-407f1800-9853-11ea-80e2-34b71a320778.png)

**Comment by snakefoot** [comment]
@amikapoor Thank you for benchmarks. A good starting point for making improvements. Could be nice with some details out the benchmark. How many logevents in total. How many logevents per second?

Your eagerness to continue change the NetworkTarget makes me think that you have not read my first reply. I think it is better to create a dedicated HttpClientTarget. Any reason you think otherwise?

**Comment by amikapoor** [comment]
@snakefoot Thanks for the reply. I read your earlier reply and my thinking is that if NLog is going to have a `HttpNetworkSender` at all, then it should at least function in a way that doesn't cause the app to break due to using so many connections. I get that the Network target works really well with the `TcpNetworkSender`, but as soon as `HttpNetworkSender` comes into play, the connections sky rocket.

For each test it was a simple for loop logging 40,000 events with no rest in between. All the logs came in within 2-3 min.

**Comment by snakefoot** [comment]
@amikapoor Well the ability to do http-requests is a "hidden" secret of the network-target (Mostly used for TCP/ UDP). Usually people go for the WebService-Target when doing http. Because it has a lot more options for configuring the http-requests (http-headers, proxy etc.)

I think it is great that you want to improve the NLog-library, and pull-requests are always welcome. Just think you might not get the best performance (which seems the goal) by giving the `HttpNetworkSender` a beating. But please prove me wrong.

**Comment by amikapoor** [comment]
@snakefoot ah got it thanks. Didn't realize it was a hidden secret. Question though, what is the reason for keeping this hidden secret around if something like a `WebService` target already exists?

I will run another test in a bit using the `WebService` target just for reference. But I think it will suffer from the same problem because looking at the code, it still uses `HttpWebRequest` instead of `HttpClient`, which I think is a big reason for these issues.

**Comment by snakefoot** [comment]
You are completely right that WebService-Target suffers from the same problem as `HttpNetworkSender`. Also WebService-Target is intended to perform web-service-requests, where a single LogEvent matches a single web-request.

And again you will notice that WebService-Target will not give optimal performance compared the loggly-target. Again see my first comment about the missing ability to batch multiple logevents into a single http-request (Something that both WebService-Target and Network-Target cannot do).

**Comment by amikapoor** [comment]
In both Loggly target tests, I did not use batching. I did 1 log statement per request. Performance was bad using the old version because connection close was specified in each request. Once that header was removed, HttpClient was able to reuse the connections so performance was good even without batching (low TCP and SNAT numbers).

Basically what I’m saying is that I think we can achieve the same thing in NLog directly just by using a single static HttpClient which is the recommended way to perform requests instead of using HttpWebRequest, even without batching. The changes could be in both HttpNetworkSender, or WebService Target, or both. I can make a PR and do more tests. The problem is HttpClient is not available on all the platforms as @304NotModified stated earlier.

**Comment by snakefoot** [comment]
HttpClient will also bring in a new dependency (system.net.http.dll), and we are working towards reducing them for NLog (Splitting into more nuget-packages).

Eventhough HttpClient is included with Net45 (and newer):

https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=netframework-4.5

Then I think here are bugfixes included in the nuget-package:

https://packages.nuget.org/packages/System.Net.Http/

So eventhough you want to "fix" HttpNetworkSender and WebService-Target, then I think the fixing should happen in a new nuget-package. Maybe this HttpClient nuget-package will include NetworkTarget and WebServiceTarget when NLog 5.0 is released.

**Comment by snakefoot** [comment]
@amikapoor Maybe you can verify that this NLog-target will also resolve your performance-issues:

https://github.com/DarekDan/NLog.Targets.HTTP (Nuget: [NLog.Targets.Http](https://www.nuget.org/packages/NLog.Targets.HTTP))
